docs/porposals/diamond.md

```diff
@@ -0,0 +1,27 @@
+# Diamond workflow support.
+
+Diamond workflow allows for a pipeline to scale out the generation of a single bundle.
+
+The diamond workflow consists of the following steps.
+
+1. Before a diamond starts, Datamon SDK allows for the creation of an unique ID for a diamond.
+   1. This can be a KSUID.
+   2. The client is expected to use the KSUID at each step.
+   3. command: ```datamon bundle split initialize --repo <repo> --message```
```

Comment:
Since this is mostly about pre-uploading objects before actually committing, why don't we use the current CLI paradigm like: `bundle upload --no-commit`, which would upload, prepare some metadata (e.g. `bundles/{bundleID}/precommited/{diamon-id}/file-list.yaml`) without updating the bundle yet, the uncommitted uploads. Each participant to the diamond could thus upload things pre-emptively, at any time, possibly even several times without committing.
Eventually, the standard `bundle upload` 

Benefit:
* Client wouldn't have to remember the KSUID: pre-emptive upload implicitly all relate to the next commit
* We don't have to break up the process in an initialize and a split stage: clients are expected to `bundle upload --no-commit`, then we only expect some minimal coordination to issue a final `bundle upload`.
* Since the command is reused, messages and labels can be specified and merged by the eventual commit
* This way of doing thing may be used to "diamond" but also as a way to resume things whenever a failure occurs during an upload. We would have to modify a bit the current upload so it always starts by creating this "staging" and would resume automatically when restarted.
* Requirements on client coordination are less stringent: we can even let clients push intermediary commits (they would be very fast), provided datamon guarantees same the shuffling of metadata is atomic (e.g. a lock file is needed to move and merge the content of `bundles/{bundleID}/precommited` to `bundles/{bundleID}`

Reply:
how would we deal with concurrency? There can be multiple split uploads on going for the same repo. We do not have a parent child relationship between bundles. I see this feature more analogous to the multi part upload of S3. Also, the bundle bucket is expected to be write once and never update to guarantee no change in data.  We can discuss this more in person.

Reply:
Concurrency would be achieved by allowing each participant to write pre-commited resources to a unique location in the config (e.g. `bundles/{bundleID}/precommited/{diamon-id}/file-list.yaml`) then putting a `.ready` signal there when it is done with its upload operation.

Since the "bundle" is not yet committed maybe the path should be more like: 
bundles/{base bundle ID|"first"}/precommit/{diamond-id}/`
"first" here means that there is no bundle created in this repo yet.

However I admit that initiating several branches concurrently is not supported by my model. I have to think again.

Reply:
Will defer this discussion.